File: master/sos/slave.cc
/**
 * @file slave.cc
 * @brief SOS slave class implementation
 */

#include <iostream>
#include <cstring>
#include <vector>
#include <list>

#include <ctime>
#include <chrono>
#include <iomanip>			// put_tiime

#include "global.h"

#include "l2.h"
#include "msg.h"
#include "resource.h"
#include "slave.h"
#include "sos.h"

namespace sos {

/**
 * @brief Reset a slave to a known state
 */

void slave::reset (void)
{
    if (addr_)
	delete addr_ ;			// XXX should be lock-protected
    l2_ = 0 ;
    addr_ = 0 ;
    reslist_.clear () ;
    status_ = SL_INACTIVE ;
    next_timeout_ = std::chrono::system_clock::time_point::max () ;
    D ("Slave " << slaveid_ << " status set to INACTIVE") ;
}

/**
 * @brief Dumps slave status and resources to an output stream
 */

std::ostream& operator<< (std::ostream &os, const slave &s)
{
    l2addr *a ;
    std::time_t nt ;
    char buf [MAXBUF] ;

    os << "slave " << s.slaveid_ << " " ;
    switch (s.status_)
    {
	case slave::SL_INACTIVE:
	    os << "INACTIVE" ;
	    break ;
	case slave::SL_RUNNING:
	    nt = std::chrono::system_clock::to_time_t (s.next_timeout_) ;
	    strftime (buf, sizeof buf, "%F %T", std::localtime (&nt)) ;
	    os << "RUNNING (mtu=" << s.mtu_ << ", ttl=" << buf << ")" ;
	    break ;
	default:
	    os << "(unknown state)" ;
	    break ;
    }
    a = ((slave *) (&s))->addr () ;	// why "a = s.addr ()" doesn't work?
    if (a)
	os << " mac=" << *a ;
    os << "\n" ;

    // Display resources
    for (auto &r : s.reslist_)
	os << r ;

    return os ;
}

/**
 * @brief Look-up the given resource
 *
 * Find a given resource /a/b/c on the slave. The resource is
 * represented as a vector of string (vector [] = { a, b, c}).
 *
 * @param v vector containing the components of the resource URL
 * @return pointer to resource, if found. NULL otherwise.
 */

resource *slave::find_resource (const std::vector <std::string> &v)
{
    resource *r = nullptr ;

    for (auto &res : reslist_)
    {
	if (res == v)
	{
	    r = &res ;
	    break ;
	}
    }
    return r ;
}

/**
 * @brief Fetches the resources advertised by the slave
 *
 * Fetches all the resources advertised by a slave in its
 * Assoc Answer message (its /.well-known/sos).
 *
 * @return vector of resource objects
 */

const std::vector <resource> &slave::resource_list (void)
{
    return reslist_ ;
}


/**
 * @brief SOS protocol handling
 *
 * This method is called by a receiver thread when the received
 * message is a control message originated from this slave.
 * The method implements the SOS control protocol for this
 * slave, and maintains the state associated to this slave.
 *
 * @param e current sos engine
 * @param m received message
 */

void slave::process_sos (sos *e, msgptr_t m)
{
    switch (m->sos_type ())
    {
	case msg::SOS_DISCOVER :
	    {
		D ("Received DISCOVER, sending ASSOCIATE") ;
		msgptr_t answer (new msg) ;
		answer->peer (m->peer ()) ;
		answer->type (msg::MT_CON) ;
		answer->code (msg::MC_POST) ;
		answer->mk_ctl_assoc (init_ttl_) ;
		e->add_request (answer) ;
	    }
	    break ;
	case msg::SOS_ASSOC_REQUEST :
	    D ("Received ASSOC_REQUEST from another master") ;
	    break ;
	case msg::SOS_ASSOC_ANSWER :
	    // has answer been correlated to a request?
	    if (m->reqrep ())
	    {
		byte *pload ; int plen ;
		std::vector <resource> rlist ;

		D ("Received ASSOC ANSWER for slave" << slaveid_) ;

		 // Add ressource list from the answer
		pload = (byte *) m->payload (&plen) ;
		D ("payload length=" << plen) ;

		if (parse_resource_list (rlist, pload, plen))
		{
		    D ("Slave " << slaveid_ << " status set to RUNNING") ;
		    status_ = SL_RUNNING ;
		    reslist_ = rlist ;
		    next_timeout_ = DATE_TIMEOUT_S (init_ttl_) ;
		}
		else
		    D ("Slave " << slaveid_ << " cannot parse resource list") ;
	    }
	    break ;
	case msg::SOS_HELLO :
	    D ("Received HELLO from another master") ;
	    break ;
	default :
	    D ("Received an unrecognized message") ;
	    break ;
    }
}

bool slave::parse_resource_list (std::vector <resource> &rlist, const byte *b, int len)
{
    enum { s_start,			// potential terminal state
	    s_resource,
	    s_endres,			// potential terminal state
	    s_attrname,			// potential terminal state
	    s_attrval_start,		// potential terminal state
	    s_attrval_nquoted,		// potential terminal state
	    s_attrval_quoted,
	    s_error
	} state = s_start ;
    std::string current ;		// string currently being built
    std::string attrname ;		// attribute name

    // Ref: RFC 6690

    // XXX : this code does not check syntax of link-param and values
    // XXX : this code does nothing with link-param and values

    for ( ; len > 0 ; len--, b++)
    {
	switch (state)
	{
	    case s_start :
		if (*b == '<')
		{
		    current = "" ;
		    state = s_resource ;
		}
		else state = s_error ;
		break ;
	    case s_resource :
		if (*b == '>')
		{
		    rlist.push_back (resource (current)) ;
		    state = s_endres ; 
		}
		else current += *b ;
		break ;
	    case s_endres :
		attrname = "" ;
		if (*b == ';')
		    state = s_attrname ;
		else if (*b == ',')
		    state = s_start ;
		else state = s_error ;
		break ;
	    case s_attrname :
		if (*b == '=')
		{
		    if (attrname == "")
			state = s_error ;
		    else
			state = s_attrval_start ;
		}
		else if (*b == ';')
		{
		    rlist.back ().add_attribute (attrname, "") ;
		    attrname = "" ;
		    state = s_attrname ;	// stay in current state
		}
		else if (*b == ',')
		{
		    rlist.back ().add_attribute (attrname, "") ;
		    state = s_start ;
		}
		else attrname += *b ;
		break ;
	    case s_attrval_start :
		current = "" ;
		if (*b == '"')
		    state = s_attrval_quoted ;
		else if (*b == ',')
		{
		    rlist.back ().add_attribute (attrname, "") ;
		    state = s_start ;
		}
		else if (*b == ';')
		{
		    rlist.back ().add_attribute (attrname, "") ;
		    attrname = "" ;
		    state = s_attrname ;
		}
		else
		{
		    state = s_attrval_nquoted ;
		    current += *b ;
		}
		break ;
	    case s_attrval_quoted :
		if (*b == '"')
		{
		    rlist.back ().add_attribute (attrname, current) ;
		    state = s_endres ;
		}
		else current += *b ;
		break ;
	    case s_attrval_nquoted :
		if (*b == ',')
		{
		    rlist.back ().add_attribute (attrname, current) ;
		    state = s_start ;
		}
		else if (*b == ';')
		{
		    rlist.back ().add_attribute (attrname, current) ;
		    attrname = "" ;
		    state = s_attrname ;
		}
		else current += *b ;
		break ;
	    default :
		state = s_error ;
		break ;
	}
    }
    
    // take care of terminal states
    switch (state)
    {
	case s_start :
	    // nothing to do
	    break ;
	case s_endres :
	    break ;
	case s_attrname :
	    if (attrname == "")
		rlist.back ().add_attribute (attrname, "") ;
	    break ;
	case s_attrval_start :
	    rlist.back ().add_attribute (attrname, "") ;
	    break ;
	case s_attrval_nquoted :
	    rlist.back ().add_attribute (attrname, current) ;
	    break ;
	default :
	    // nothing to do
	    state = s_error ;
	    break ;
    }

    return state != s_error ;
}

}					// end of namespace sos

File: pySos/sos/slave.py
"""
This module contains the Slave class
"""

from datetime import datetime, timedelta
from enum import Enum
from util.debug import *
from sos import msg
from .resource import Resource

class Slave:
    """
    This class describes a slave in the SOS system.
    """
    class StatusCodes(Enum):
        SL_INACTIVE = 0
        SL_ASSOCIATING = 0
        SL_RUNNING = 1

    class ResStatus(Enum):
        S_START = 0
        S_RESOURCE = 1
        S_ENDRES = 2
        S_ATTRNAME = 3
        S_ATTRVAL_START  = 4
        S_ATTRVAL_NQUOTED = 5
        S_ATTRVAL_QUOTED = 6
        S_ERROR = 7

    def __init__(self):
        """
        Default constructor
        """
        self.id = 0
        self.def_mtu = 0
        self.curmtu = 0
        self.l2_net = None
        self.addr = None
        self.init_ttl = 0
        self.status = self.StatusCodes.SL_INACTIVE
        self.res_list = []
        self.next_timeout = datetime.now()

    def __str__(self):
        """
        Returns a string describing the slave status and resources.
        """
        s1 = 'slave ' + str(self.id) + ' '
        s2 = ('INACTIVE' if self.status is self.StatusCodes.SL_INACTIVE else
              'RUNNING (curmtu=' + str(self.curmtu) + ', ttl= ' +
              self.next_timeout.isoformat() + ')'
              if self.status is self.StatusCodes.SL_RUNNING else
              '(unknown state)')
        s3 = ' mac=' + str(self.addr) + '\n'
        return s1 + s2 + s3

    def reset(self):
        """
        Resets the slave to it's default state
        """
        self.addr = None
        self.l2_net = None
        self.res_list.clear()
        self.curmtu = 0
        self.status = self.StatusCodes.SL_INACTIVE
        self.next_timeout = datetime.now()
        print_debug(dbg_levels.STATE, 'Slave' + str(self.id) + 
                    ' status set to INACTIVE.')

    def find_resource(self, res):
        """
        Looks for a given resource on the slave.
        This function expects the resource to be an iterable object
        containing the resource path, e.g for resource '/a/b/c' : ['a','b','c']
        If the resource is not found, returns None
        """
        for r in self.res_list:
            if r == res:
                return r
        return None

    def resource_list(self):
        """
        Returns the list of all the resources managed by the slave
        """
        return self.res_list

    def process_sos(self, sos_instance, mess):
        """
        This method is called by a receiver thread when the received message
        is a control message originated from this slave.
        The method implements the SOS control protocol for this slave, 
        and maintains the state associated to this slave.
        """
        tp = mess.sos_type
        if tp is msg.Msg.SosTypes.SOS_DISCOVER:
            print_debug(dbg_levels.STATE, 'Received DISCOVER, sending ASSOCIATE')
            answer = msg.Msg()
            answer.peer = mess.peer
            answer.msg_type = msg.Msg.MsgTypes.MT_CON
            answer.msg_code = msg.Msg.MsgCodes.MC_POST
            answer.mk_ctrl_assoc(self.init_ttl, self.curmtu)

            sos_instance.add_request(answer)

        elif tp is msg.Msg.SosTypes.SOS_ASSOC_REQUEST:
            print_debug(dbg_levels.STATE, 'Received ASSOC_REQUEST from another master')
        elif tp is msg.Msg.SosTypes.SOS_ASSOC_ANSWER:
            if mess.req_rep is not None:
                print_debug(dbg_levels.STATE, 'Received ASSOC ANSWER for slave.')
                print_debug(dbg_levels.STATE, 'payload length=' + str(len(mess.payload)))
                reslist = []
                if self.parse_resource_list(reslist, mess.payload):
                    print_debug(dbg_levels.STATE, 'Slave ' + str(self.id) + ' status set to RUNNING')
                    self.status = self.StatusCodes.SL_RUNNING
                    self.res_list = reslist
                    self.next_timeout = datetime.now() + timedelta(seconds=self.init_ttl)
                else:
                    print_debug(dbg_levels.STATE, 'Slave ' + str(self.id) + ' cannot parse resource list.')
        elif tp is msg.Msg.SosTypes.SOS_HELLO:
            print_debug(dbg_levels.STATE, 'Received HELLO from another master')
        else:
            print_debug(dbg_levels.STATE, 'Received an unrecognized message')

    def parse_resource_list(self, rlist, payload):
        attrname = ''
        cur_res = bytearray()
        state = None
        def parse_single_byte(state, b):
            nonlocal rlist, attrname, cur_res
            if state is self.ResStatus.S_START:
                if b == b'<':
                    cur_res = bytearray()
                    return self.ResStatus.S_RESOURCE
                else:
                    return self.ResStatus.S_ERROR
            elif state is self.ResStatus.S_RESOURCE:
                if b == b'>':
                    rlist.append(Resource(cur_res.decode()))
                    return self.ResStatus.S_ENDRES
                else:
                    cur_res += b
                    return state
            elif state is self.ResStatus.S_ENDRES:
                attrname = ''
                if b == b';':
                    return self.ResStatus.S_ATTRNAME
                elif b == b',':
                    return self.ResStatus.S_START
                else:
                    return self.ResStatus.S_ERROR
            elif state is self.ResStatus.S_ATTRNAME:
                if b == b'=':
                    if attrname == '':
                        return self.ResStatus.S_ERROR
                    else:
                        return self.ResStatus.S_ATTRVAL_START
                elif b == b';':
                    rlist[len(rlist)-1].add_attribute(attrname, '')
                    attrname = ''
                    return state
                elif b == b',':
                    rlist[len(rlist)-1].add_attribute(attrname, '')
                    return self.ResStatus.S_START
                else:
                    attrname = attrname + b.decode()
                    return state
            elif state is self.ResStatus.S_ATTRVAL_START:
                cur_res = bytearray()
                if b == b'"':
                    return self.ResStatus.S_ATTRVAL_QUOTED
                elif b == b',':
                    rlist[len(rlist)-1].add_attribute(attrname, '')
                    return self.ResStatus.S_START
                elif b == b';':
                    rlist[len(rlist)-1].add_attribute(attrname, '')
                    attrname = ''
                    return self.ResStatus.S_ATTRNAME
                else:
                    cur_res += b
                    return self.ResStatus.S_ATTRVAL_NQUOTED
            elif state is self.ResStatus.S_ATTRVAL_QUOTED:
                if b == b'"':
                    rlist[len(rlist)-1].add_attribute(attrname, cur_res.decode())
                    return self.ResStatus.S_ENDRES
                else:
                    cur_res += b
                    return state
            elif state is self.ResStatus.S_ATTRVAL_NQUOTED:
                if b == b',':
                    rlist[len(rlist)-1].add_attribute(attrname, cur_res.decode())
                    return self.ResStatus.S_START
                elif b == b';':
                    rlist[len(rlist)-1].add_attribute(attrname, cur_res.decode())
                    attrname = ''
                    return self.ResStatus.S_ATTRNAME
                else:
                    cur_res += b
            else:
                return self.ResStatus.S_ERROR

        state = self.ResStatus.S_START
        # Iterating on a bytes object yields integers but parse_single_byte takes bytes.
        # So, iterate on indices and give it slices of length 1.
        for i in range(len(payload)):
            state = parse_single_byte(state, payload[i:i+1])
            if state is self.ResStatus.S_ERROR:
                break

        # Handle terminal states
        if state is self.ResStatus.S_ATTRNAME:
            if attrname == '':
                rlist[len(rlist)-1].add_attribute(attrname, '')
        elif state is self.ResStatus.S_ATTRVAL_START:
            rlist[len(rlist)-1].add_attribute(attrname, '')
        elif state is self.ResStatus.S_ATTRVAL_NQUOTED:
            rlist[len(rlist)-1].add_attribute(attrname, cur_res.decode())
        elif state is self.ResStatus.S_ENDRES:
            pass
        else:
            state = self.ResStatus.S_ERROR
        return state is not self.ResStatus.S_ERROR
